vsscale/generic.py

```diff
@@ -0,0 +1,152 @@
+from __future__ import annotations
+
+from abc import ABC, abstractmethod
+from typing import Any, Protocol
+
+from jetpytools import inject_self
+
+from vskernels import Catrom, Kernel, KernelT, Scaler, ScalerT
+from vstools import ConstantFormatVideoNode, MatrixT, check_variable, plane, vs
+
+__all__ = [
+    'BaseGenericScaler',
+    'GenericScaler',
+]
+
+
+class _GeneriScaleNoShift(Protocol):
+    def __call__(
+        self, clip: ConstantFormatVideoNode, width: int, height: int,
+        *args: Any, **kwargs: Any
+    ) -> ConstantFormatVideoNode:
+        ...
+
+
+class _GeneriScaleWithShift(Protocol):
+    def __call__(
+        self, clip: ConstantFormatVideoNode, width: int, height: int, shift: tuple[float, float],
+        *args: Any, **kwargs: Any
+    ) -> ConstantFormatVideoNode:
+        ...
+
+
+def _func_no_op(clip: ConstantFormatVideoNode, *args: Any, **kwargs: Any) -> ConstantFormatVideoNode:
+    return clip
+
+
+class BaseGenericScaler(Scaler, ABC):
+    """
+    Generic Scaler base class.
+    Inherit from this to create more complex scalers with built-in utils.
+    """
+
+    def __init__(
+        self,
+        *,
+        kernel: KernelT = Catrom,
+        scaler: ScalerT | None = None,
+        shifter: KernelT | None = None,
+        **kwargs: Any
+    ) -> None:
+        """
+        :param kernel:      Base kernel to be used for certain scaling/shifting/resampling operations.
+                            Defaults to Catrom.
+        :param scaler:      Scaler used for scaling operations. Defaults to kernel.
+        :param shifter:     Kernel used for shifting operations. Defaults to scaler.
+        """
+        self.kernel = Kernel.ensure_obj(kernel, self.__class__)
+        self.scaler = Scaler.ensure_obj(scaler or self.kernel, self.__class__)
+        self.shifter = Kernel.ensure_obj(
+            shifter or (self.scaler if isinstance(self.scaler, Kernel) else Catrom), self.__class__
```

Comment:
would it be a bad idea to fallback to `self.kernel`?

```suggestion
            shifter or (self.scaler if isinstance(self.scaler, Kernel) else self.kernel), self.__class__
```

Reply:
I’m not against it. Makes the code cleaner as well.

Reply:
I misunderstood. I thought you meant : 

```py
self.shifter = Kernel.ensure_obj(shifter or self.kernel, self.__class__)
```